kvdb/cache.go

```diff
@@ -0,0 +1,930 @@
+}
+
+func (c *cacheReadTx) ForEachBucket(f func(key []byte) error) error {
+	return forEachBucketImpl(c.cache, f)
+}
+
+func (c *cacheReadTx) Rollback() error {
+	if c.active {
+		defer func() {
+			c.active = false
+			c.cache.mx.RUnlock()
+		}()
+
+		return c.dbTx.Rollback()
+	}
+
+	return nil
+}
+
+type cacheReadWriteTx struct {
+	cache    *Cache
+	dbTx     walletdb.ReadWriteTx
+	active   bool
+	onCommit func()
+}
+
+func newCacheReadWriteTx(cache *Cache,
+	dbTx walletdb.ReadWriteTx) *cacheReadWriteTx {
+
+	return &cacheReadWriteTx{
+		cache:  cache,
+		dbTx:   dbTx,
+		active: true,
+	}
+}
+
+func (c *cacheReadWriteTx) ReadBucket(key []byte) walletdb.ReadBucket {
+	return readBucketImpl(c.cache, c.dbTx, key)
+}
+
+func (c *cacheReadWriteTx) ForEachBucket(f func(key []byte) error) error {
+	return forEachBucketImpl(c.cache, f)
+}
+
+func (c *cacheReadWriteTx) Rollback() error {
+	if c.active {
+		defer func() {
+			c.active = false
+			c.cache.mx.Unlock()
+		}()
+
+		// First revert changes to the cache itself.
+		for e := c.cache.pending.Front(); e != nil; e = e.Next() {
+			e.Value.(pendingChange).Revert()
+		}
+
+		// Now that we got back the old cache state, we can reset the
+		// pending change list and revert the DB transaction.
+		c.cache.pending = list.New()
+		return c.dbTx.Rollback()
+	}
+
+	return nil
+}
+
+func (c *cacheReadWriteTx) ReadWriteBucket(key []byte) walletdb.ReadWriteBucket {
+	cacheItem := c.cache.root.bucket.get(key)
+	if cacheItem == nil {
+		return nil
+	}
+
+	dbBucket := c.dbTx.ReadWriteBucket(key)
+	if dbBucket == nil {
+		return nil
+	}
+
+	return newCacheReadWriteBucket(c.cache, cacheItem, dbBucket)
+}
+
+func (c *cacheReadWriteTx) CreateTopLevelBucket(key []byte) (
+	walletdb.ReadWriteBucket, error) {
+
+	dbBucket, err := c.dbTx.CreateTopLevelBucket(key)
+	if err != nil {
+		return nil, err
+	}
+
+	bucketRoot := c.cache.root.bucket.get(key)
+	if bucketRoot != nil {
+		return newCacheReadWriteBucket(
+			c.cache, bucketRoot, dbBucket,
+		), nil
+	}
+
+	bucketRoot = &cachedItem{
+		key:    string(key),
+		bucket: newCacheBucket(),
+	}
+
+	c.cache.root.bucket.put(bucketRoot)
+	c.cache.pendingAdd(c.cache.root, bucketRoot)
+
+	return newCacheReadWriteBucket(c.cache, bucketRoot, dbBucket), nil
+}
+
+func (c *cacheReadWriteTx) DeleteTopLevelBucket(key []byte) error {
+	if err := c.dbTx.DeleteTopLevelBucket(key); err != nil {
+		return err
+	}
+
+	return deleteFromCache(c.cache, c.cache.root, key, true)
+}
+
+func (c *cacheReadWriteTx) Commit() error {
+	if err := c.dbTx.Commit(); err != nil {
+		return err
+	}
+
+	defer func() {
+		c.active = false
+		c.cache.mx.Unlock()
+	}()
+
+	c.cache.pending = list.New()
+	if c.onCommit != nil {
+		c.onCommit()
+	}
+
+	return nil
+}
+
+func (c *cacheReadWriteTx) OnCommit(f func()) {
+	c.onCommit = f
+}
+
+func forEachImpl(root *cachedItem, f func(k, v []byte) error) error {
+	var err error
+
+	root.bucket.tree.Ascend(func(item btree.Item) bool {
```

Comment:
No need to bubble up the error (which is `false`) here?

Reply:
The error is set below: `if err = f([]byte(c.key), val); err != nil {` and then returned at the end. So if the tree iterator returns with false, we set the error defined outside the closure and just return that.